### PR TITLE
Sdil 1746 vary while returning

### PR DESCRIPTION
**Pull request by mattrobertsky**
Please review.

### CODE REVIEW

File: app/sdil/controllers/ReturnsController.scala [hunk 1]
```diff
@@ -197,21 +216,82 @@ class ReturnsController (
...
+        packingSites.isEmpty
+      packingSites     <- askPackSites(packingSites ++ firstPackingSite.fold(List.empty[Site])(x => List(x)))
+    } yield packingSites
+  }
+
   private def program(period: ReturnPeriod, subscription: Subscription, sdilRef: String)
                      (implicit hc: HeaderCarrier): WebMonad[Result] = for {
     sdilReturn     <- askReturn(subscription, sdilRef)
+    // check if they need to vary
+    isNewImporter   = (sdilReturn.importLarge |+| sdilReturn.importSmall) > 0 && !subscription.activity.importer
```
Comment (hmrclt) [inline app/sdil/controllers/ReturnsController.scala]: consider using nonEmpty and doing away with LongComparison

File: app/sdil/controllers/ReturnsController.scala [hunk 2]
```diff
@@ -197,21 +216,82 @@ class ReturnsController (
...
+    inner           = uniform.fragments.return_variation_continue(isNewImporter, isNewPacker)
+    _               <- tell("return-change-registration", inner) when isNewImporter || isNewPacker
+    newPackingSites <- askNewPackingSites(subscription) when isNewPacker && subscription.productionSites.isEmpty
+    newWarehouses   <- askNewWarehouses when isNewImporter && subscription.warehouseSites.isEmpty
+
+    variation     = ReturnsVariation(
+      orgName = subscription.orgName,
+      ppobAddress = subscription.address,
+      importer = (isNewImporter, (sdilReturn.importLarge |+| sdilReturn.importSmall).combineN(4)),
+      packer = (isNewPacker, (sdilReturn.packLarge |+| sdilReturn.packSmall.total).combineN(4)),
```
Comment (hmrclt) [inline app/sdil/controllers/ReturnsController.scala]: maybe consider having `totalPacked` and `totalImported` methods on sdilReturn if you're using them a lot?

File: app/sdil/controllers/ReturnsController.scala [hunk 3]
```diff
@@ -197,21 +216,82 @@ class ReturnsController (
...
     broughtForward <- BigDecimal("0").pure[WebMonad]
-    _              <- checkYourAnswers("check-your-answers", sdilReturn, broughtForward, subscription.activity.smallProducer)
+    _              <- checkYourAnswers("check-your-answers", sdilReturn, broughtForward, variation, subscription)
     _              <- cachedFuture(s"return-${period.count}")(
                         sdilConnector.returns(subscription.utr, period) = sdilReturn)
+    _              <- if (isNewImporter || isNewPacker) {
+                        execute(sdilConnector.returns.variation(variation, sdilRef))
+                      } else {
+                        (()).pure[WebMonad]
+                      }
```
Comment (hmrclt) [inline app/sdil/controllers/ReturnsController.scala]: Do you think we need a uniform construct for "if (a) then do b and throw away the result"?

File: app/sdil/models/ReturnsVariation.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+package sdil.models
+
+import play.api.libs.json.{Format, Json}
+import sdil.models.backend.{Site, UkAddress}
+
+case class ReturnsVariation(
+  orgName: String,
+  ppobAddress: UkAddress,
+  importer: (Boolean, (Long,Long)) = (false, (0,0)),
```
Comment (hmrclt) [inline app/sdil/models/ReturnsVariation.scala]: what would (true,(0,0)) mean? or (false,(1,1))?

File: app/sdil/controllers/ReturnsController.scala (hunk 2 above)
Comment (mattrobertsky) [reply app/sdil/controllers/ReturnsController.scala]: Hmm, not sure, should the variation data not be distinct from the return data, even if it is derived from it?

File: app/sdil/controllers/ReturnsController.scala (hunk 3 above)
Comment (mattrobertsky) [reply app/sdil/controllers/ReturnsController.scala]: Yes please.

File: app/sdil/models/ReturnsVariation.scala (hunk 4 above)
Comment (mattrobertsky) [reply app/sdil/models/ReturnsVariation.scala]: true (0,0) would be an anomoly, but false (1,1) would not. 

I guess both packer and importer could be set to (0,0) if they were not new but I thought it was clearer in the code to explicitly say they are new importer/copacker and the vols are..

Comment (hmrclt) [reply app/sdil/models/ReturnsVariation.scala]: ah I had assumed the boolean referred to if there was any activity, not any change.

**Pull request merged**